### ISSUE TITLE
Incorrectly believe that Tron is an immediate loss

### ISSUE THREAD

**Issue by arr28** [SanchoGGP/ggp-base#270]
In [this match](http://www.ggp.org/view/tiltyard/matches/dff54d62e395c123c22b87aefd0400dd6069aa2a/) of Tron, we commit suicide.  That's because we think that all 4 moves lead to an immediate loss.

Logs available on Tiltyard.

**Comment by arr28** [comment]
This has happened in at least two more matches.

**Comment by arr28** [comment]
Reliably repros using the `TiltyardTest`.  (Even if it happens to pick the correct move, the logs clearly show a complete loss at depth 1.)

The first time I ran it, the "most likely path" contained a large number (~10) of sensible moves, but it still thought that moving right was a complete loss in 3.  The second time I ran it, it thought that right was immediately a complete loss (depth 1).

**Comment by arr28** [comment]
Okay, so I suspect this is an issue with completion in simultaneous move games.  If we play with a small enough node pool or time limit that the root doesn't complete, we always select the only play which isn't immediately fatal.

If running for long enough that the root does complete, I see various behaviour, but all seem to be correctly acknowledging that everything except `right` leads to an immediate loss (or draw if the opponent messes up).  The difference is then in the what happens when playing `right`.  This tends to be reported as a complete loss at various depths, but all of them very shallow (1-3).  This isn't right.  A lower bound on the number of moves we can definitely survive is 10.

In a sample where `right` claimed to complete at depth 2, the (37MB) tree dump showed...

```
Root scores [0.00, 100.00] (complete@26)
  @1: Move up [0] D23 (choosing role 1) scores [0.00, 100.00] (complete@24)[0.0@0.0] (ref 8589948001) - visits: 85 (85, false), updates: 88.0
    @2: Move left [0] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8589974948) - visits: 19 (19, false), updates: 0.0
    @2: Move down [1] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8589974952) - visits: 45 (45, false), updates: 0.0
    @2: Move up [2] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8589974951) - visits: 19 (19, false), updates: 0.0
    @2: Move right [3] D24 (choosing role 0) scores [50.00, 50.00]T (complete@24)[0.0@0.0] (ref 8589974950) - visits: 1 (1, false), updates: 0.0
  @1: Move left [1] D23 (choosing role 1) scores [0.00, 100.00] (complete@24)[0.0@0.0] (ref 8589934804) - visits: 88 (88, false), updates: 91.0
    @2: Move left [0] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8589934972) - visits: 19 (19, false), updates: 0.0
    @2: Move down [1] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8589939100) - visits: 47 (47, false), updates: 0.0
    @2: Move up [2] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8589945144) - visits: 20 (20, false), updates: 0.0
    @2: Move right [3] D24 (choosing role 0) scores [50.00, 50.00]T (complete@24)[0.0@0.0] (ref 8589953917) - visits: 1 (1, false), updates: 0.0
  @1: Move down [3] D23 (choosing role 1) scores [0.00, 100.00] (complete@24)[0.0@0.0] (ref 8589940291) - visits: 88 (88, false), updates: 91.0
    @2: Move left [0] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8589940290) - visits: 19 (19, false), updates: 0.0
    @2: Move down [1] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8590046794) - visits: 47 (47, false), updates: 0.0
    @2: Move up [2] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8590046796) - visits: 20 (20, false), updates: 0.0
    @2: Move right [3] D24 (choosing role 0) scores [50.00, 50.00]T (complete@24)[0.0@0.0] (ref 8590046797) - visits: 1 (1, false), updates: 0.0
  @1: Move right [2] D23 (choosing role 1) scores [0.00, 100.00] (complete@26)[0.0@0.0] (ref 8589974949) - visits: 482816 (482816, false), updates: 1715170.0
  ...
```

...which all seems to be okay.  If we go up/left/down, the game is immediately terminal and we either score 0 or 50 (if the other player commits suicide by doubling back).

**Comment by arr28** [comment]
I've examined a bit of the "what happens when we move `right`?" branch.  It looks basically sound, but the completion depths are all wrong.  (Note that in the tree dump, they're expressed as tree depth from the original root, whereas on the console, they're reported as number of turns from the current position.)

There's 1 branch in the dump which is expanded to a further depth of 28 (14 moves) below the current position.  Presumably this is how long we can hang on for given best play by our opponent.

```
 @1: Move right [2] D23 (choosing role 1) scores [0.00, 100.00] (complete@26)[0.0@0.0] (ref 8589974949) - visits: 482816 (482816, false), updates: 1715170.0
 @2: Move up    [2] D24 (choosing role 0) scores [52.13, 47.87][0.0@0.0] (ref 8589957330) - visits: 57522 (57522, false), updates: 209252.0
 @3: Move down  [3] D25 (choosing role 1) scores [52.13, 47.87][0.0@0.0] (ref 8589957584) - visits: 57518 (57518, false), updates: 209244.0
 @4: Move right [3] D26 (choosing role 0) scores [54.17, 45.83][0.0@0.0] (ref 8590067913) - visits: 11489 (11489, false), updates: 40903.0
 @5: Move down  [3] D27 (choosing role 1) scores [54.29, 45.71][0.0@0.0] (ref 8590067916) - visits: 11385 (11385, false), updates: 40786.0
 @6: Move right [3] D28 (choosing role 0) scores [58.43, 41.57][0.0@0.0] (ref 17180332211) - visits: 3581 (3581, false), updates: 12931.0
 @7: Move left  [1] D29 (choosing role 1) scores [59.55, 40.45][0.0@0.0] (ref 25769806672) - visits: 3295 (3295, false), updates: 12150.0
 @8: Move right [3] D30 (choosing role 0) scores [62.06, 37.94][0.0@0.0] (ref 17180387317) - visits: 1954 (1954, false), updates: 7224.0
 @9: Move down  [3] D31 (choosing role 1) scores [64.51, 35.49][0.0@0.0] (ref 17180045743) - visits: 1386 (1386, false), updates: 5103.0
@10: Move up    [2] D32 (choosing role 0) scores [64.33, 35.67][0.0@0.0] (ref 25769807355) - visits: 1385 (1385, false), updates: 5098.0
@11: Move left  [1] D33 (choosing role 1) scores [64.49, 35.51][0.0@0.0] (ref 17180230059) - visits: 1369 (1369, false), updates: 5038.0
@12: Move up    [2] D34 (choosing role 0) scores [67.58, 32.42][0.0@0.0] (ref 17180159154) - visits: 751 (751, false), updates: 2655.0
@13: Move up    [0] D35 (choosing role 1) scores [70.94, 29.06][0.0@0.0] (ref 17180647810) - visits: 579 (579, false), updates: 2033.0
@14: Move left  [0] D36 (choosing role 0) scores [56.27, 43.73][0.0@0.0] (ref 17180872424) - visits: 315 (315, false), updates: 1136.0
@15: Move left  [1] D37 (choosing role 1) scores [56.53, 43.47][0.0@0.0] (ref 34360649845) - visits: 309 (309, false), updates: 1114.0
@16: Move down  [1] D38 (choosing role 0) scores [55.36, 44.64][0.0@0.0] (ref 34359789924) - visits: 156 (156, false), updates: 583.0
@17: Move left  [1] D39 (choosing role 1) scores [59.19, 40.81][0.0@0.0] (ref 17180490913) - visits: 121 (121, false), updates: 465.0
@18: Move left  [0] D40 (choosing role 0) scores [58.84, 41.16][0.0@0.0] (ref 17180642353) - visits: 120 (120, false), updates: 460.0
@19: Move left  [1] D41 (choosing role 1) scores [64.03, 35.97][0.0@0.0] (ref 17180837636) - visits: 64 (64, false), updates: 260.0
@20: Move left  [0] D42 (choosing role 0) scores [58.56, 41.44][0.0@0.0] (ref 25770677488) - visits: 40 (40, false), updates: 164.0
@21: Move down  [3] D43 (choosing role 1) scores [54.11, 45.89][0.0@0.0] (ref 25770120953) - visits: 12 (12, false), updates: 60.0
@22: Move left  [0] D44 (choosing role 0) scores [44.95, 55.05][0.0@0.0] (ref 25769854303) - visits: 8 (8, false), updates: 40.0
@23: Move left  [1] D45 (choosing role 1) scores [43.35, 56.65][0.0@0.0] (ref 17180187023) - visits: 6 (6, false), updates: 30.0
@24: Move left  [0] D46 (choosing role 0) scores [45.96, 54.04][0.0@0.0] (ref 17180205977) - visits: 5 (5, false), updates: 25.0
@25: Move left  [1] D47 (choosing role 1) scores [44.96, 55.04][0.0@0.0] (ref 17180205978) - visits: 4 (4, false), updates: 20.0
@26: Move down  [1] D48 (choosing role 0) scores [24.92, 75.08][0.0@0.0] (ref 17180139536) - visits: 2 (2, false), updates: 10.0
@27: Move up    [0] D49 (choosing role 1) scores [10.00, 90.00][0.0@0.0] (ref 25769817359) - visits: 1 (1, false), updates: 5.0
@28: Move up    [2] D50 (choosing role 0) scores [100.00, 0.00]T (complete@50)[0.0@0.0] (ref 25769818133) - visits: 0 (0, false), updates: 0.0
```

**Comment by arr28** [comment]
Okay, so the branch above is pretty unilluminating, but I now reckon that 14 is a lower bound on the number of moves we can survive in that position.  (Actually, I think it's 16, because it's in our interests to play chicken when we meet, but red should refuse to play chicken an just seal us off, leaving the extra 2 moves.)

BTW, all of the above is for [this match](http://www.ggp.org/view/tiltyard/matches/dff54d62e395c123c22b87aefd0400dd6069aa2a/) (the link in the opening issue statement).

**Comment by arr28** [comment]
Okay - I now have a tree dump from an instance where we thought that all moves were immediate losses (i.e. complete at depth 1).  I'll take a look.

**Comment by arr28** [comment]
Found this in `MCTSTree.processNodeCompletions()`...

```java
      //  Rarely we can have a situation where a complete node is queued, but gets freed by a parent completion DIRECTLY
      //  processed from expand, and then reallocated before it gets processed here.  To address this flaw we really ought
      //  to queue refs rather than actual nodes, but it is extremely rare, so, for now we just (heuristically) trap it
      //  by checking this is actually a compete unfreed node
      if (!node.freed && node.complete)
      {
        node.processCompletion();
      }
```

There were a couple of long periods in the game while our node table was completely full.  Mostly that was at the start, but the second was 5 moves before the end.  At that point, I can imagine we were completing a significant number of branches.  Perhaps we're being bitten by this?  (Not convinced though and will look for other explanations first.)

**Comment by SteveDraper** [comment]
Maybe worth a bit of refactoring to queue refs and eliminate this as both a
possibility here, and a potential future bug course?

**Comment by arr28** [comment]
I've done the refactoring, but it didn't fix the problem - as expected given the additional digging I've done.

I'm stepping back a bit here and trying to define what it ought to mean for a node to be "complete" in a simultaneous game.

If you're able to comment on what you think we try to do at the moment, that would be great.  (This might not be the same as what I come up with, but it would be good to know what the current intention is.)

**Comment by SteveDraper** [comment]
I believe that we require all (or one decisive) children to be complete in
the ply that starts the joint move (i.e. - although we don;t know the other
roles' simul moves in that ply we still complete if it doesn;t matter, so
that ply is essentially the same as non-simul).  In other plys (the ones
where the choice is feeding into a joint move which already has decisions
of other players, that are not known, built into the node parentage) we
only complete if all cousins are complete for the decisive case (so a move
we would choose is complete in all branches and wins in all branches), or
all same-level-as-child nodes are complete across the entire joint move for
the non-decisive case.  In the decisive case we calculate a 'floor' score
(worst possible outcome across the unknown opponent moves) and consider
that a decisive completion only if the floor is above (some threshold
condition I forget).  In the everything-complete so propagate completion
case we blend the scores across all possible joint moves in the proportion
of the scores of the choosers for the other parts of the joint move (i.e. -
most probable blend in some sense)

**Comment by SteveDraper** [comment]
Just a thought on the refactoring to use (Long) refs rather than direct TreeNode references - I noticed in diffing this that the queue is (and always was) a linked list.  Both adding list nodes and boxing the refs as Longs will allocate memory.  I can't really see any reason not to turn this into a circular array of 'long' with some reasonable max size (just sleep when trying to queue if said max is ever exceeded, but in reality this rarely goes above single figures, so making the max O(256) or so should mean that almost never occurs)

**Comment by arr28** [comment]
Good idea.  I'll do something along those lines.

So here's one of the invalid trees, reporting fully to depth 4 below the current root (which is at depth 22).  Presumably this ought to be enough to determine (a) whether the root should be marked complete yet and (b) what depth it should be showing (or at least why it shouldn't be showing 24).

```
Root scores [0.00, 100.00] (complete@24)
  @1: Move left [0] D23 (choosing role 1) scores [0.00, 100.00] (complete@24)[0.0@0.0] (ref 8590002599) - visits: 75 (75, false), updates: 83.0
    @2: Move up [0] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8590002598) - visits: 17 (17, false), updates: 0.0
    @2: Move right [1] D24 (choosing role 0) scores [50.00, 50.00]T (complete@24)[0.0@0.0] (ref 8590002597) - visits: 1 (1, false), updates: 0.0
    @2: Move left [2] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 17180521061) - visits: 17 (17, false), updates: 0.0
    @2: Move down [3] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 17180521059) - visits: 39 (39, false), updates: 0.0
  @1: Move right [1] D23 (choosing role 1) scores [0.00, 100.00] (complete@24)[0.0@0.0] (ref 8590590274) - visits: 934875 (934875, false), updates: 1.1236808E7
    @2: Move up [0] D24 (choosing role 0) scores [52.06, 47.94][0.0@0.0] (ref 17180475699) - visits: 366911 (366911, false), updates: 4312327.0
      @3: Move left [0] D25 (choosing role 1) scores [0.00, 100.00] (complete@26)[0.0@0.0] (ref 17180348980) - visits: 10 (10, false), updates: 0.0
        @4: Move up [0] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 17180348979) - visits: 0 (0, false), updates: 0.0
        @4: Move right [1] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 17180348968) - visits: 0 (0, false), updates: 0.0
        @4: Move left [2] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 17180348972) - visits: 0 (0, false), updates: 0.0
        @4: Move down [3] D26 (choosing role 0) scores [50.00, 50.00]T (complete@26)[0.0@0.0] (ref 17180348970) - visits: 0 (0, false), updates: 0.0
      @3: Move right [1] D25 (choosing role 1) scores [0.00, 100.00] (complete@26)[0.0@0.0] (ref 17180348986) - visits: 10 (10, false), updates: 0.0
        @4: Move up [0] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 17180348987) - visits: 0 (0, false), updates: 0.0
        @4: Move right [1] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 17180348973) - visits: 0 (0, false), updates: 0.0
        @4: Move left [2] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 17180348974) - visits: 0 (0, false), updates: 0.0
        @4: Move down [3] D26 (choosing role 0) scores [50.00, 50.00]T (complete@26)[0.0@0.0] (ref 17180348975) - visits: 0 (0, false), updates: 0.0
      @3: Move down [2] D25 (choosing role 1) scores [52.06, 47.94][0.0@0.0] (ref 8590367292) - visits: 366907 (366907, false), updates: 4312318.0
        @4: Move up [0] D26 (choosing role 0) scores [51.42, 48.58][0.0@0.0] (ref 8590327393) - visits: 244009 (244009, false), updates: 2844705.0
        @4: Move right [1] D26 (choosing role 0) scores [53.98, 46.02][0.0@0.0] (ref 17180348969) - visits: 46739 (46739, false), updates: 558393.0
        @4: Move left [2] D26 (choosing role 0) scores [52.96, 47.04][0.0@0.0] (ref 8590837137) - visits: 76158 (76158, false), updates: 909211.0
        @4: Move down [3] D26 (choosing role 0) scores [100.00, 0.00]T (complete@26)[0.0@0.0] (ref 8590837103) - visits: 11 (11, false), updates: 0.0
      @3: Move up [3] D25 (choosing role 1) scores [0.00, 100.00] (complete@26)[0.0@0.0] (ref 17180348976) - visits: 10 (10, false), updates: 0.0
        @4: Move up [0] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 17180348977) - visits: 0 (0, false), updates: 0.0
        @4: Move right [1] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 17180348978) - visits: 0 (0, false), updates: 0.0
        @4: Move left [2] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 17180348982) - visits: 0 (0, false), updates: 0.0
        @4: Move down [3] D26 (choosing role 0) scores [50.00, 50.00]T (complete@26)[0.0@0.0] (ref 17180348981) - visits: 0 (0, false), updates: 0.0
    @2: Move right [1] D24 (choosing role 0) scores [100.00, 0.00]T (complete@24)[0.0@0.0] (ref 8590590273) - visits: 12 (12, false), updates: 0.0
    @2: Move left [2] D24 (choosing role 0) scores [53.34, 46.66][0.0@0.0] (ref 8590589111) - visits: 114016 (114016, false), updates: 1376460.0
      @3: Move left [0] D25 (choosing role 1) scores [7.90, 92.10][0.0@0.0] (ref 8590334009) - visits: 48 (48, false), updates: 56.0
        @4: Move up [0] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 8590334008) - visits: 15 (15, false), updates: 0.0
        @4: Move right [1] D26 (choosing role 0) scores [50.00, 50.00]T (complete@26)[0.0@0.0] (ref 8590334007) - visits: 1 (1, false), updates: 0.0
        @4: Move left [2] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 8590334006) - visits: 16 (16, false), updates: 0.0
        @4: Move down [3] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 17179911795) - visits: 15 (15, false), updates: 0.0
      @3: Move right [1] D25 (choosing role 1) scores [7.90, 92.10][0.0@0.0] (ref 8589983656) - visits: 48 (48, false), updates: 56.0
        @4: Move up [0] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 8589983655) - visits: 15 (15, false), updates: 0.0
        @4: Move right [1] D26 (choosing role 0) scores [50.00, 50.00]T (complete@26)[0.0@0.0] (ref 8589983654) - visits: 1 (1, false), updates: 0.0
        @4: Move left [2] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 8589983653) - visits: 16 (16, false), updates: 0.0
        @4: Move down [3] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 8590334005) - visits: 15 (15, false), updates: 0.0
      @3: Move down [2] D25 (choosing role 1) scores [53.36, 46.64][0.0@0.0] (ref 8590367295) - visits: 113883 (113883, false), updates: 1376295.0
        @4: Move up [0] D26 (choosing role 0) scores [53.61, 46.39][0.0@0.0] (ref 17179880501) - visits: 34647 (34647, false), updates: 417974.0
        @4: Move right [1] D26 (choosing role 0) scores [100.00, 0.00]T (complete@26)[0.0@0.0] (ref 8590367291) - visits: 10 (10, false), updates: 0.0
        @4: Move left [2] D26 (choosing role 0) scores [53.15, 46.85][0.0@0.0] (ref 34360582614) - visits: 41512 (41512, false), updates: 494139.0
        @4: Move down [3] D26 (choosing role 0) scores [53.35, 46.65][0.0@0.0] (ref 8590779022) - visits: 37723 (37723, false), updates: 464173.0
      @3: Move up [3] D25 (choosing role 1) scores [1.13, 98.87][0.0@0.0] (ref 17180624999) - visits: 36 (36, false), updates: 44.0
        @4: Move up [0] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 17180625000) - visits: 11 (11, false), updates: 0.0
        @4: Move right [1] D26 (choosing role 0) scores [50.00, 50.00]T (complete@26)[0.0@0.0] (ref 17180635868) - visits: 1 (1, false), updates: 0.0
        @4: Move left [2] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 17180635869) - visits: 12 (12, false), updates: 0.0
        @4: Move down [3] D26 (choosing role 0) scores [0.00, 100.00]T (complete@26)[0.0@0.0] (ref 8589983652) - visits: 11 (11, false), updates: 0.0
    @2: Move down [3] D24 (choosing role 0) scores [0.00, 100.00] (complete@26)[0.0@0.0] (ref 25769818356) - visits: 453947 (453947, false), updates: 5548012.0
  @1: Move down [2] D23 (choosing role 1) scores [0.00, 100.00] (complete@24)[0.0@0.0] (ref 8590367352) - visits: 82 (82, false), updates: 90.0
    @2: Move up [0] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8590367354) - visits: 19 (19, false), updates: 0.0
    @2: Move right [1] D24 (choosing role 0) scores [50.00, 50.00]T (complete@24)[0.0@0.0] (ref 8590590270) - visits: 1 (1, false), updates: 0.0
    @2: Move left [2] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8590590271) - visits: 18 (18, false), updates: 0.0
    @2: Move down [3] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8590590275) - visits: 43 (43, false), updates: 0.0
  @1: Move up [3] D23 (choosing role 1) scores [0.00, 100.00] (complete@24)[0.0@0.0] (ref 8590590260) - visits: 82 (82, false), updates: 90.0
    @2: Move up [0] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8590590264) - visits: 19 (19, false), updates: 0.0
    @2: Move right [1] D24 (choosing role 0) scores [50.00, 50.00]T (complete@24)[0.0@0.0] (ref 8590590263) - visits: 1 (1, false), updates: 0.0
    @2: Move left [2] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8590590262) - visits: 18 (18, false), updates: 0.0
    @2: Move down [3] D24 (choosing role 0) scores [0.00, 100.00]T (complete@24)[0.0@0.0] (ref 8590590261) - visits: 43 (43, false), updates: 0.0
```

**Comment by arr28** [comment]
Looking just at the immediate responses to use moving right...

```
  @1: Move right [1] D23 (choosing role 1) scores [0.00, 100.00] (complete@24)[0.0@0.0] (ref 8590590274) - visits: 934875 (934875, false), updates: 1.1236808E7
    @2: Move up [0] D24 (choosing role 0) scores [52.06, 47.94][0.0@0.0] (ref 17180475699) - visits: 366911 (366911, false), updates: 4312327.0
    @2: Move right [1] D24 (choosing role 0) scores [100.00, 0.00]T (complete@24)[0.0@0.0] (ref 8590590273) - visits: 12 (12, false), updates: 0.0
    @2: Move left [2] D24 (choosing role 0) scores [53.34, 46.66][0.0@0.0] (ref 8590589111) - visits: 114016 (114016, false), updates: 1376460.0
    @2: Move down [3] D24 (choosing role 0) scores [0.00, 100.00] (complete@26)[0.0@0.0] (ref 25769818356) - visits: 453947 (453947, false), updates: 5548012.0
```

...one of the 4 is a complete & immediate win for us (if our opponent commits suicide by going back on himself - i.e right), 2 are incomplete and 1 is (supposedly) a complete loss for us, but at later depth 26 (which clearly isn't true, it's probably just the same bug at lower scope, but let's pretend for now that it is).

Given those 4 children, I claim that it's wrong for us moving right to be complete at 24.  I think it would be correct (ignoring the bug's effect further down the tree) to mark this as a complete loss at depth 26.

Steve - do you agree with my analysis?

**Comment by SteveDraper** [comment]
I certainly agree that we should not be claiming its completion depth is 24, given that we know some paths don;t complete for 26.  However, it may well not be correct to mark it as a complete loss at 26 just because one opponent move has that effect.  In simultaneous move games it can often be the case that every option we have is a complete loss for SOME opponent move (consider blocker - any move for the crosser is a complete loss in sequences where the blocker presciently plays the same move all the time).  This is why we need all cousins to also be complete at this ply to conclude that our move is complete (that is all possible opponent simultaneous choices lead to a complete result), and furthermore why even then the value has to be a weighted blend (because the opponent is actually not prescient)

**Comment by arr28** [comment]
Yes, but the point here is that (according to the dump) if our opponent plays `down`, it's a complete win for him whatever we choose to do.  He doesn't need to be prescient.

**Comment by SteveDraper** [comment]
AH yes, that's true - I hadn't looked at all the siblings (i.e. - that down is a win from all our move choices)

**Comment by arr28** [comment]
This looks a bit better...

```
22:08:19.296 - Move right scores -450.00 (selectionScore score -449.81, selection count 631008, ref 25771484531, complete [25])
22:08:19.297 - Move left scores -498.00 (selectionScore score -497.79, selection count 151, ref 25769813339, complete [1])
22:08:19.297 - Move down scores -498.00 (selectionScore score -497.79, selection count 159, ref 34361519296, complete [1])
22:08:19.297 - Move up scores -498.00 (selectionScore score -497.79, selection count 168, ref 60130757278, complete [1])
```

Left/Up/Down are all immediately complete losses.  Right isn't a complete loss until much later.  (This is with 120s deadlines, so ~4mins thinking time.)

However, I'm still not convinced that it's quite right.  On best play by our opponent, I don't think we can hang on for 25 moves.

**Comment by SteveDraper** [comment]
Don't forget that complete[25] means 25 node layers, which is only 12.5
moves in a 2 player game.  Does that make things more plausible?

**Comment by arr28** [comment]
Sadly not.  Number of players is corrected for at this point in the output...

```
((child.mCompletionDepth - mTree.mRoot.mDepth)/mTree.mNumRoles)
```

**Comment by arr28** [comment]
I'll retest in the light of https://github.com/SanchoGGP/ggp-base/commit/1d0945a2b59fa4dbe4655bcf3164cebe68e6d7df.

**Comment by SteveDraper** [comment]
Ok - suggest you review what I did there.  The thing to note (but I don;t
think it arises in Tron) is that it is possible to transpose to the same
state at differing depths in some games.  Hence a child's depth can be
lower (or the same) as its parent's.  When such a child was completed the
parent completion depth could thus be <= its own depth.  What the change
does is calculate the relative completion depth (completionDepth-depth) of
the completing node and completes the parent at <parent depth>+<complete
child relative completion depth>+1.  Note that the +1 part is itself only a
constraining approximation (since non-decision nodes could have been
arbitrarily removed), but should preserve at least a correct ordering.

**Comment by arr28** [comment]
We can't possibly reach the same state at different depths in Tron because, on each turn, exactly 2 more cells are filled in (or the game becomes terminal).

**Comment by SteveDraper** [comment]
Exactly, so I don't think this has any implications in Tron.

**Comment by SteveDraper** [comment]
Found a couple of minor completion depth issues, but generally speaking I think this is fine now.  The completion depth is an upper bound not a lower bound, since propagation of complete children will mean that the first solution found (not necessarily the most efficient) can cause the deletion of sibling paths.  The order things are found in is somewhat non-deterministic, but the nature of completion is such that it implies there is no way out for the opponent (but not that its the fastest way to win).  Hence a figure of 25 here (I observed 23 and 25 in testing on various runs) just means that the winner can guarantee a win using a conservative strategy that allows up to 25 moves before the win finishes.

Closing this along with #277, having fixed some incorrect completion depths found from adding assertions when investigating #277 (one particularly bad one of which was only simultaneous move games)
